docs/proposal-new-output-syntax.md

```diff
@@ -0,0 +1,702 @@
+leads at runtime.
+
+Cylc 7 could tell (albeit imperfectly!) if the intended path was followed 
+because all task outputs were effectively *required*. All possible paths are
+pre-spawned (to the next cycle point instance of each active task, roughly
+speaking). Users had to explicitly remove unused alternate paths with suicide
+triggers, and any remaining unsatisfied waiting tasks would indicate a problem
+by stalling the workflow.
+
+This document proposes a way to convey the intended path of execution to the
+Cylc 8 spawn-on-demand scheduler by making all task outputs **expected** by
+default and giving a new syntax to explicitly mark **optional** outputs. If a
+task finishes without completing all of its expected outputs it will be
+retained in the n=0 task pool and flagged as an **incomplete task** (if the
+task did not fail unexpectedly then this represents an error against the
+workflow configuration that was not recognized by the task job itself).
+
+*Note: optional triggers were considered and rejected as another way of
+potentially handling this problem - see cylc-admin#128*
+
+## The New Output Syntax
+
+`foo:x` means `x` is an *expected output* of task `foo`. If a task finishes
+without completing all of its expected outputs, we will retain it in the task
+pool as an *incomplete task*.
+
+`foo?y` means `y` is an *optional output* of task `foo`. It is OK for a task to
+finish without completing its optional outputs.
+
+Abbreviated syntax for the success case:
+- `foo?` means `foo?succeed` (success is optional)
+- `foo` and `foo:` mean `foo:succeed` (success is expected)
+  - this supports Cylc 7 graphs as well as being a sensible default
+
+## Syntax Errors
+
+### succeed and fail
+
+The `succeed` and `fail` outputs are mutually exclusive opposites, therefore:
+
+- `succeed` and `fail` can't both be expected (or the task would be flagged as
+  incomplete no matter what)
+```
+# ERROR
+a
+a:fail
+```
+
+- `succeed` and `fail` can both be optional though (at a success/fail alternate
+  branching point)
+```
+# OK
+a?
+a?fail
+```
+
+- however it is OK to have one of `succeed` and `fail` expected and one
+  optional. This allows triggering off of the not-expected output at the same
+  time as the task is flagged as incomplete
+```
+# OK
+a
+a?fail => call_the_police
+```
+(Note this means we can't have *"`a` means `a:succeed` unless declared optional
+elsewhere"*.)
+
+### Custom Outputs
+
+Custom outputs must be either expected or optional, they can't be both
+```
+# ERROR
+a:x
+a?x
+```
+
+### QUESTION: submit-fail, submit, and start?
+
+It makes some sense to allow the `fail` output to be expected, but allowing
+`submit-fail` to be expected means treating a task as incomplete if it
+successfully submits. The only use case I've thought of
+is to make it somewhat easier (less cleanup before shutdown) to write
+functional tests that deliberately submit to a fake platform.
```

Comment:
Can live without it. I'm not sure that its enough of a use case.

Reply:
Agreed.

Reply:
Does it really make sense to allow `fail` to be expected? Retry is designed around the fact that tasks are expected to succeed. If a task which is expected to fail succeeds do we need to be able to configure it to retry until it fails? That seems crazy. I think I'd prefer us to say that `fail` (and `submit-fail`) cannot be expected.

Reply:
Fair enough. I suppose if a user really has an executable that they expect not to exit with `0` whilst still doing what they need it to do, they can provide a simple wrapper to turn that into success for the workflow.